enterprise/tools/replay_action/replay_action.go

```diff
@@ -0,0 +1,440 @@
+			log.Fatalf("Error copying command: %s", err.Error())
+		}
+		treeRN := digest.NewResourceName(action.GetInputRootDigest(), *sourceRemoteInstanceName, rspb.CacheType_CAS, repb.DigestFunction_SHA256)
+		tree, err := cachetools.GetTreeFromRootDirectoryDigest(srcCtx, sourceCASClient, treeRN)
+		if err != nil {
+			log.Fatalf("GetTree failed: %s", err.Error())
+		}
+		if err := copyTree(rootCtx, fmb, destBSClient, sourceBSClient, tree); err != nil {
+			log.Fatalf("Error copying files: %s", err.Error())
+		}
+		log.Infof("Finished copying files.")
+	}
+	execReq := &repb.ExecuteRequest{
+		InstanceName:    *targetRemoteInstanceName,
+		SkipCacheLookup: true,
+		ActionDigest:    d,
+	}
+	eg := &errgroup.Group{}
+	eg.SetLimit(*jobs)
+	for i := 1; i <= *n; i++ {
+		i := i
+		eg.Go(func() error {
+			execute(targetCtx, execClient, destBSClient, i, d, execReq)
+			return nil
+		})
+	}
+	eg.Wait()
+}
+
+func execute(ctx context.Context, execClient repb.ExecutionClient, bsClient bspb.ByteStreamClient, i int, d *repb.Digest, req *repb.ExecuteRequest) {
+	log.Infof("Starting action %d of %d...", i, *n)
+	stream, err := execClient.Execute(ctx, req)
+	if err != nil {
+		log.Fatalf(err.Error())
+	}
+	printedExecutionID := false
+	for {
+		op, err := stream.Recv()
+		if err != nil {
+			log.Fatalf("Error on stream: %s", err.Error())
+		}
+		if !printedExecutionID {
+			log.Infof("Started task %q", op.GetName())
+			printedExecutionID = true
+		}
+		log.Infof("Execution stage: %s", operation.ExtractStage(op))
+		if op.GetDone() {
+			metadata := &repb.ExecuteOperationMetadata{}
+			if err := op.GetMetadata().UnmarshalTo(metadata); err == nil {
+				jb, _ := (protojson.MarshalOptions{Multiline: true}).Marshal(metadata)
+				log.Infof("Metadata: %s", string(jb))
+			}
+
+			response := &repb.ExecuteResponse{}
+			if err := op.GetResponse().UnmarshalTo(response); err != nil {
+				log.Errorf("Failed to unmarshal response: %s", err)
+				return
+			}
+
+			if err := gstatus.ErrorProto(response.GetStatus()); err != nil {
+				log.Errorf("Execution failed: %s", err)
+				break
+			}
+
+			jb, _ := (protojson.MarshalOptions{Multiline: true}).Marshal(response)
+			log.Infof("ExecuteResponse: %s", string(jb))
+			result := response.GetResult()
+			if result.GetExitCode() != 0 {
+				log.Warningf("Action exited with code %d", result.GetExitCode())
+			}
+			// Print stdout and stderr but only when running a single action.
+			if *n == 1 {
+				printOutputFile(ctx, bsClient, result.GetStdoutDigest(), "stdout")
+				printOutputFile(ctx, bsClient, result.GetStderrDigest(), "stderr")
+			}
+			logExecutionMetadata(i, d, response.GetResult().GetExecutionMetadata())
+			break
+		}
+	}
+}
+
+type findMissingRequest struct {
+	Digest       *repb.Digest
+	ResponseChan chan findMissingResponse
+}
+
+type findMissingResponse struct {
+	Missing bool
+	Error   error
+}
+
+type FindMissingBatcherOpts struct {
```

Comment:
I don't think this should be required for replay_action to work, should it? Would it be sufficient to just send single find missing requests?

Reply:
For actions with lots of inputs (which are often the ones I want to replay) the round-trip overhead of the FindMissing requests is very high. For example I tried replaying one of our heavier test actions and when fully cached it takes 20s for the FindMissing checks (without batching), vs ~2s with batching enabled.

The goal here was mainly just to save time & frustration when trying to debug - I would use the `-skip_copy_files` flag to save time but it was frustrating because I would keep forgetting to set it.

Reply:
Sgtm